### PR TITLE
Add support for python-can

### LINKED ISSUE #2: use python-can as CAN layer?
Hi,

Do you know python-can. I am wondering if it is possible to use that for the CAN layer.

https://python-can.readthedocs.io

cheers

### PR DESCRIPTION
**Pull request by christiansandberg**
This pull request adds support for many CAN interfaces and operating systems through the python-can package + a basic ISO-TP implementation. I realise this library should be transport agnostic but since the majority of UDS implementations are using CAN and ISO-TP, I'm hoping this will be OK to include anyway. Otherwise the isotp module could perhaps be a separate package.

I've got some more testing to do but I'm opening this now for discussion.

Fixes #2.

### CODE REVIEW

**Comment by pylessard** [comment]
Hi Christian,
Thank you for this. I will review the code as soon as I can and let you know.
Regards

**Comment by pylessard** [comment]
Hi Christian,
Great works!
Do you think it would be better to handle the reception of ISOTP frame in a separate thread? The problem I see with the actual implementation is that if wait_frame() is not called with the right timing, the sender may trigger a timeout and stop sending.

File: udsoncan/connections.py [hunk 1]
```diff
@@ -291,6 +298,74 @@ def empty_rxqueue(self):
...
+
+	def empty_rxqueue(self):
+		while self.bus.recv(0) is not None:
+			pass
+
+	def specific_send(self, payload):
+		self.isotp_send(payload)
+
+	def specific_wait_frame(self, timeout=2):
+		return self.isotp_recv(timeout)
```
Comment (pylessard) [inline udsoncan/connections.py]: An exception should be raised if connection is not opened

File: udsoncan/connections.py [hunk 2]
```diff
@@ -291,6 +298,74 @@ def empty_rxqueue(self):
...
+
+	def is_open(self):
+		return self.bus is not None
+
+	def empty_rxqueue(self):
+		while self.bus.recv(0) is not None:
+			pass
+
+	def specific_send(self, payload):
+		self.isotp_send(payload)
```
Comment (pylessard) [inline udsoncan/connections.py]: n exception should be raised if connection is not opened. Sending without opening the conenction raise "None has no attribute sends()", which is misleading

**Comment by christiansandberg** [comment]
I don't believe it will be a problem, especially since the library does not support asynchronous communication anyway. Usually you will send a request then immediately wait for a response. The peer should wait up to 1 second before timing out so that should not be a problem. One use-case I can see being a problem is sending multiple requests on many connections then reading the responses one by one.

If we do need a thread, I need to rewrite the receiving of messages so they can be directed to either the receive thread or an ongoing send operation. One alternative is using my [aioisotp](https://github.com/christiansandberg/aioisotp) package (under development) and run the asyncio event loop in a thread to allow synchronous operation.

**Comment by pylessard** [comment]
Correct, we're not doing any asynchronous exchange and I don't plan to add support for this soon. Still, I believe having the reception in a thread will make the software more robust against race condition, or simply less restrictive for the programmer. Imagine you send a request, then your OS switch task, it may take 16-32ms on a windows machine. An embedded device largely have time to receive the request, start responding and timeout.  Also, it is hard to manually test the connection without this :)

What I suggest is just to move the receive logic into a thread, and put the received data into a queue, then wait_frame() reads that queue. Actually, reception and transmit logic could go in that task. That would imply very few modification to the code and will add robustness timing-wise.

SocketCAN works like this, you can sends many frames into the socket, each frames are queued on the other side until they are read.

Last thing, I tested the IsoTP connection against harktopp/can-isotp module through a vcan interface. That worked, but when I slowed down the transmission I started hitting a strange issue. Actually, try setting stmin to 50ms. 4 times out of 5, your stack complains about a missing flow control message, but candump shows that it was there. It seems that python-can receive nothing. I haven't investigated too much yet.

**Comment by christiansandberg** [comment]
I thought it would be best to keep the ISO-TP implementation in a separate package. That would keep udsoncan free from the more complex transport implementation. It should support full duplex transfers. First I have to update python-can and aioisotp and get them released on PyPI.

**Comment by pylessard** [comment]
@christiansandberg How do we stand on this improvement? I won't merge anything unless you tell me it is ready.

**Comment by christiansandberg** [comment]
I think it is implemented, so it should be possible to install and test it out. But I wanted to test it out myself first but haven't had much time.

```
$ pip install 
https://github.com/christiansandberg/python-udsoncan/archive/python-can-support.zip
```

**Comment by pylessard** [comment]
I'll give it a try then

**Comment by christiansandberg** [comment]
Sorry for the late reply. I can see the issue. A temporary workaround is probably to add a short delay before the first send() call.

**Comment by syntroniks** [comment]
I found this to be working with an ICSNeo device with the following packages:
```
aioisotp==0.1.1
python-can==3.0.0
python-ics==2.15
udsoncan==1.2 (installed from https://github.com/christiansandberg/python-udsoncan/archive/python-can-support.zip)
```

```
import udsoncan
from udsoncan.connections import PythonCanConnection

conn = PythonCanConnection(rxid=0x7E9, txid=0x7E1, channel=1, interface='neovi')
conn.open()
conn.send(b'\x3E\x00')
payload = conn.wait_frame(timeout=1, exception=True)
print(payload)
```
This provides a better interface than my PR #13 and I would love to be able to use it with UDS messages.

**Comment by christiansandberg** [comment]
I did some tweak that hopefully fixed the issue. But broadening the scope of this library to also include an ISO-TP implementation could be better.

**Comment by pylessard** [comment]
Will look soon. As I mentioned to @syntroniks on #13 , I started a Python implemntation of IsoTP since python-can is more and more requested.
I was planning to include it in my isotp package and then add a Connection that would use the module.

I still think that we could include aioisotp in udsoncan, maybe I'll rename the connection so we can make the difference between both implementations.  Maybe PythonCanAsyncIsotpConnection

Any thoughts on this?

**Comment by christiansandberg** [comment]
No need to keep both. The original idea of aioisotp was to be used standalone in an asyncio application.

**Comment by pylessard** [comment]
Ok, then I'll keep this PR open until I finish my works. I plan this before end of 2018.
Cheers

**Pull request closed without merging**